---
title: "sSNAPPY: an R/Bioconductor package for single-sample directional pathway perturbation analysis"
author: 
  - name: Wenjun Liu
    affiliation: Dame Roma Mitchell Cancer Research Laboratories, Adelaide Medical School, Faculty of Health and Medical Sciences, University of Adelaide, Adelaide, Australia
    email: Corresponding Author wenjun.liu@adelaide.edu.au
  - name: Ville-Petteri Mäkinen
    affiliation: 
    - Computational Medicine, Faculty of Medicine, University of Oulu, Oulu, Finland
    - Center for Life Course Health Research, Faculty of Medicine, University of Oulu, Oulu, Finland
  - name: Wayne D. Tilley
    affiliation: Dame Roma Mitchell Cancer Research Laboratories, Adelaide Medical School, Faculty of Health and Medical Sciences, University of Adelaide, Adelaide, Australia
  - name: Stephen M. Pederson
    affiliation: 
    - Dame Roma Mitchell Cancer Research Laboratories, Adelaide Medical School, Faculty of Health and Medical Sciences, University of Adelaide, Adelaide, Australia
    - Black Ochre Data Laboratories, Telethon Kids Institute, Adelaide, Australia
    - John Curtin School of Medical Research, Australian National University, Canberra, Australia
abstract: 
 A common outcome of analysing RNA-Seq data is the detection of biological pathways with significantly altered activity between the conditions under investigation. 
 Whilst many strategies test for over-representation within pre-defined gene-sets for genes showing changed expression, these analyses typically do not account for gene-gene interactions encoded by pathway topologies, and are not able to directly predict the directional change of pathway activity. 
 To address these issues, we have deloped a single-sample pathway perturbation analysis method  [*sSNAPPY*](https://bioconductor.org/packages/sSNAPPY), now available as an R/Bioconductor package, which leverages pathway topology information to compute pathway perturbation scores, and predicts the direction of change across a set of pathways. 
 Here, we demonstrate the use of *sSNAPPY* by applying the method to public scRNA-seq data, derived from ovarian cancer patient tissues collected before and after chemotherapy. 
 Not only were we able to predict the directions of significant perturbations of pathways discussed in the original study, but *sSNAPPY* was also able to detect significant changes of other biological processes, yielding far greater insight into the response to treatment.
  *sSNAPPY* represents a novel pathway analysis strategy that takes into consideration of pathway topology to predict impacted biology pathways, both within related samples and across treatment groups. 
  In addition to not relying on the detection of differentially expressed genes, the method and associated R package offer important flexibility and provide powerful visualisation tools.

keywords: RNA-seq, pathway enrichment, R package, topology, KEGG, scRNA-seq
bibliography: sample.bib
output: BiocWorkflowTools::f1000_article

header-includes:
  - \usepackage{endfloat}    
  - \usepackage{booktabs}
  - \usepackage{longtable}
---

**R version**: `r R.version.string`

**Bioconductor version**: `r BiocManager::version()`

**Package**: `r packageVersion("sSNAPPY")` 

# Introduction

Using pathway enrichment analysis to gain biological insights from gene expression data is a pivotal step in the analysis and interpretation of RNA-seq data, for which numerous methods have been developed (reviewed in [@Maleki2020-ur; @Mubeen2022-eq]). 
Many existing methods tend to view pathways simply as a collection of gene names, as seen in those relying on the detection of differentially expressed genes and applying over-representation analysis (ORA) strategies, and those scoring all genes using functional class scoring (FCS), such as in Gene Set Enrichment Analysis (GSEA) [@Subramanian2005-lx], arguably the most widely-used approach. 
However, databases such as the Kyoto Encyclopaedia of Genes and Genomes (KEGG)[@OgataKEGGKyotoEncyclopediaa] and WikiPathways[@Martens2021] capture not only which genes are implicated in a certain biological process but also their interactions, activating or inhibitory roles, and their relative importance within the pathway, all of which are overlooked in ORA- and FCS-based approaches. 

To fully utilise that additional information, the latest generation of pathway analysis approaches include many which are topology-based such as SPIA[@Tarca2009], DEGraph[@Jacob2012], NetGSA[@Ma2016] and PRS[@Ibrahim2012], as well as others which explicitly model inter-gene correlations[@Wu2012].
Despite differences in the null hypotheses tested across these approaches, overall, they have demonstrated enhanced sensitivity and specificity due to their abilities to take gene-gene interconnections into account[@Nguyen2019-va; @Ma2019]. 
Nevertheless, most topology-based methods focus only on comparing activities of pathways between two treatment groups and cannot be used to score individual samples (Figure \@ref(fig:Figure1)). 
However, in heterogenous data where more than one factor may be influencing observations[@Hanzelmann2013], incorporating scoring within paired samples may be desirable and may be able to reveal more nuanced insights. 
To address this gap, we present a Single-Sample directioNAl Pathway Perturbation analYsis methodology called [*sSNAPPY*](https://bioconductor.org/packages/sSNAPPY), available as an R/Bioconductor package.
This article defines how *sSNAPPY* computes changes in gene expression within paired samples, and propagates this through gene-set topologies to predict the perturbation in pathway activities within paired samples, before providing summarised results across an entire dataset (Figure \@ref(fig:Figure1)). 
The practical usage of the *sSNAPPY* R/Bioconductor package is illustrated through the analysis of a public scRNA-seq dataset using pseudo-bulk strategies.

```{r Figure1, out.width='100%',fig.cap="Schematic illustration of the differences betweeen conventional pathway analysis methods and sSNAPPY. Instead of being limited to treatment-level analyses, \\textit{sSNAPPY} allows the detection of pathway perturbation in individual samples by using sample-specific estimates of fold-change instead of experiment-wide estimates. (Created with BioRender.com).", echo=FALSE}
knitr::include_graphics(
    here::here("sSNAPPY_paper_files/figure-latex/Figure1.pdf")
)
```

# Methods

## Implementation

*sSNAPPY* is an R package that has been reviewed and published on the open-source bioinformatics software platform [Bioconductor](https://bioconductor.org/packages/release/bioc/html/*sSNAPPY*.html) with all source code available via GitHub.
The methodology itself is topology-based, designed to compute directional, single-sample, pathway perturbation scores in gene expression datasets with a matched-pair, or nested design (eg. samples collected before and after treatment).
This allows for the detection of pathway perturbations within all samples from a treatment group, but also within individual samples.

To run *sSNAPPY*, the only required data is a log-transformed expression matrix (e.g. logCPM) with matching sample metadata describing treatment groups and the nested structure.
It is assumed that all pre-processing has been performed beforehand, such as the exclusion of low-signal genes or normalisation to minimise technical artefacts like GC-bias.
The first step, performed internally by *sSNAPPY*, is to estimate sample-specific log fold-change ($\delta_{ghi} = \mu_{ghi} - \mu_{g0i}$) for a treatment $h$ across all genes $g$ within each patient $i$, by subtracting expression estimates for the baseline samples $\mu_{g0i}$ from those in the treatment group $h$.
Here, we gave an example of nesting samples by patients, but *sSNAPPY* could be applied to any data with matched-pair design. 
For example, RNA-seq data of treated and untreated cells from several passages, in which case the parameter $i$ would denote the passages.
It should also be noted that *sSNAPPY* supports any number of treatment/condition levels and the numbers of samples in each treatment group do not have to be identical. 
As long as a treated sample is matched to a corresponding control, the sample-specific log fold-changes will be computed for that sample.

Since it has been shown that in RNA-seq data, genes with lower expression tend to have larger variance and larger estimates of change[@Law2014], we utilise a gene-level weighting strategy to de-emphasise logFC estimates for low-abundance genes.
Gene-level weights $w_g$ are obtained in a treatment-agnostic manner by fitting a loess curve through the relationship between observed gene-level variance ($\sigma_g$) and average signal ($\bar\mu_{g}$)(Figure \@ref(fig:Figure2)), and taking the inverse of the loess-predicted variance as the weight $w_g = a / f(\bar\mu_{g})$, where $f(\bar\mu_{g})$ is the predicted value from the loess curve and the constant $a$ ensures $\sum w_g = 1$.
We then use these weighted estimates of logFC ($\delta_{ghi}^* = w_g\delta_{ghi}$) in the calculation of all pathway perturbation scores.

```{r Figure2, out.width='100%',fig.cap="Gene-wise standard deviations are plotted against the mean logCPM values with mean-variance trend modelled by a loess fit. Genes with low expression values tend to have a larger variance.", echo=FALSE}
knitr::include_graphics(
    here::here("sSNAPPY_paper_files/figure-latex/Figure2.pdf")
)
```

*sSNAPPY* was built upon the group-level topology-based scoring algorithm initially proposed in  SPIA[@Tarca2009] to propagate genes' changes in expression through pathway topologies to compute a perturbation score for each pathway.
In contrast to SPIA, which requires pre-selection of differentially expressed genes, *sSNAPPY* utilises the full expression matrix. 
By modifying the algorithm to incorporate single-sample, weighted estimates of changes in expression we are able to quantify changes in a pathway within a given sample, and then model these across all samples within a treatment group.
Thus, we define the single-sample perturbation score ($S_{hip}$) for a given pathway $p$, patient $i$ and treatment $h$:


$$
\begin{aligned}
S_{hip} = \sum_{g \in G_p} \lbrack S_{ghip} - \delta_{ghi}^*\rbrack \text{, where} \\
S_{ghip} = \delta_{ghi}^* + \sum_{g' \in U_{gp}} \beta_{gg'p} \frac{S_{g'hip}}{N_{g'p}} 
\end{aligned}
$$
where:

* $G_p$ represents the set of genes in pathway $p$, such that $g \in G_p$
* $S_{ghip}$ is the gene-, treatment- and patient-specific perturbation score for pathway $p$
* $\delta_{ghi}^* = w_g\delta_{ghi}$ is the weighted logFC of gene $g$ as described above
* $U_{gp}$ is the subset of $G_p$ containing only the genes directly upstream of gene $g$
* $\beta_{gg'p}$ is the pair-wise gene-gene interactions[@Tarca2009] encoded by the topology matrix for genes $g$ and $g'$
* $N_{gp}$ is the number of downstream genes from any gene $g$
* $S_{hip}$ is the accumulated pathway perturbation score for pathway $p$ in treatment $h$ for patient $i$

The Bioconductor package graphite[@Sales2012] provides functions that can be used to retrieve pathway topologies from a database and convert topology information to adjacency matrices.
In order to streamline this process we have implemented a convenience function, where users only need to provide the name of the desired database and species to retrieve all topology information in the format required by the scoring algorithm with the correct type of gene identifiers (ie. Entrez ID).

To scale the single-sample pathway perturbation scores ($S_{hip}$) so they are comparable across pathways and to test for significance of individual scores, null distributions of perturbation scores for each pathway are generated through a sample permutation strategy, retaining the correct correlation structure between genes within a pathway.
In the permutation, all sample labels will be randomly shuffled and assigned to permuted pairs. 
Subsequently, permuted single-sample logFCs will be calculated for each pair of permuted samples while the rest of the scoring algorithm remains unchanged.
The median and median absolute deviation (MAD) of the permuted perturbation scores will be calculated within each pathway and used to normalise the raw perturbation scores to robust $Z$-scores, which is defined as: 


$$
Z_{hip} = \frac{S_{hip} - median_{n \in N}({S_{np}})}{MAD_{n \in N}({S_{np}})}
$$
where $n$ represents each permuted pairs of samples. 
All permuted pairs possible will be constructed in this step by default unless otherwise specified. 
In a data with $I$ total samples, the maximum number of unique permuted pairs $N$ is ${}^IP_2 = \frac{{I!}}{{(I-2)!}} = I \times (I-1)$. 


Following the computation of robust $Z$-scores, two-sided *p*-values are derived to assess whether a pathway is significantly perturbed at sample-specific level.
When the sample size is small, the p-values derived from individual robust z-scores should be interpreted with caution. 
The MAD and median derived from the permuted scores are more likely to be accurately estimated when the number of permuted scores computed for each pathway is above 100, requiring at least 11 samples. 

Apart from assessing whether a pathway's activity changed significantly within an individual sample, users may also be interested in detecting changes at the group-level, which can be performed by modelling scores with regression models, incorporating Smyth's moderated $t$-statistic[@Smyth_2004] as implemented in *limma*[@limma_2015].
The single-sample nature of *sSNAPPY*'s pathway perturbation scores is particularly helpful for datasets with complex experimental designs or known confounding factors as these can also be incorporated into the final regression models.


## Operation
The package has been tested on all operating systems, requiring R > 4.3.0, and can be installed using BiocManager as follows. 

```{r eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
BiocManager::install("sSNAPPY")
```


# Use Cases

## Data

To demonstrate the application of *sSNAPPY*, we used pre-processed counts from a publicly available scRNA-seq dataset, retrieved from Gene Expression Omnibus (GEO) with accession code GSE165897. 
This dataset consists of 11 high-grade serous ovarian cancer (HGSOC) patients samples taken before and after chemotherapy[@Zhang2022].
*sSNAPPY* was used to re-analyse data from the epithelial cells as they were the primary focus of the original study. 
Since *sSNAPPY* was designed for bulk RNA-seq data, counts of epithelial cells from the same samples were first summed into pseudo-bulk profiles, giving rise to a total of 22 samples. 
We considered a gene detectable if we observed >1.5 counts per million in >11 samples out of 22, representing all samples from a complete treatment group. 
11,101 (33.8%) of the 32,847 annotated genes passed this selection criteria and were included for downstream analyses. 
Conditional quantile normalisation[@Hansen2012] was then applied to mitigate potential biases introduced by gene length and GC content. 
The normalised logCPM matrix of the processed dataset and sample metadata can be downloaded from [here](https://github.com/Wenjun-Liu/F1000_sSNAPPY_manuscript/tree/master/data).

The following packages are required for this workflow

```{r, message=FALSE}
library(sSNAPPY)
library(tidyverse)
library(magrittr)
library(ggplot2)
library(patchwork)
library(kableExtra)
library(AnnotationHub) 
library(edgeR)
library(patchwork)
library(colorspace)
```

Before running the *sSNAPPY* workflow, we need to load our expression matrix and define our sample-level metadata, then download any pathway information.
Firstly we can read in the data, setting the treatment column in the metadata to be a factor.
Importantly, to run *sSNAPPY*, the row names of the expression matrix must be specified as Entrez IDs, for compatibility with pathway databases. 
Genes without Entrez IDs were excluded, leaving 10,098 genes in the example expression matrix. 

```{r, message=FALSE}
logCPM <- read_tsv(here::here("data/logCPM.tsv")) %>%
    column_to_rownames("entrezid")
sample_meta <- read_tsv(here::here("data/sample_meta.tsv"), col_types = "cfccncnc")
head(sample_meta)
```

```{r, echo=FALSE}
# # Code for plotting figure 2 - sd vs mean relationship
# Figure_2 <- logCPM %>%
#     mutate(
#         sd = apply(., 1, sd),
#         mean = apply(., 1, mean)
#         ) %>%
#     # rownames_to_column("entrezid") %>%
#     # pivot_longer(
#     #     cols = -c("entrezid", "sd"),
#     #     names_to = "sample",
#     #     values_to = "logCPM"
#     # ) %>%
#     ggplot(
#         aes(mean, sd)
#     ) +
#     geom_point() +
#     geom_smooth(
#         method = "loess") +
#     labs(
#         x = expression(Gene-wise~average~expression~(bar(mu[g]))),
#         y = expression(Gene-wise~standard~deviation~(sigma[g]))
#     ) +
#     theme_bw() +
#     theme(
#         panel.grid = element_blank(),
#         axis.title = element_text(size = 14)
#     )
# pdf(
#     here::here("sSNAPPY_paper_files/figure-latex/Figure2.pdf") ,
#     width = 8,
#     height = 6
#     )
# Figure_2
# dev.off()
```


## Retrieval of Pathway Topology

Next, pathway topology information needs to be retrieved from a chosen database.
Using KEGG as an example, the retrieved topology information will be stored as a list where each element corresponds to a pathway and the numbers in the matrices encode gene-gene interaction. 

```{r}
gsTopology <- retrieve_topology(database = "kegg", species = "hsapiens")
```

Instead of downloading the topology matrices of all pathways, it is also possible to provide a restricted set of keywords for a targeted analysis. 
In the following we'll use the keywords 'metabolism' and 'estrogen' to only return a subset of pathways.

```{r dont-run, eval=FALSE}
# Only retrieve the topology matrices of metabolism- or signalling-related pathways
gsTopology_sub <- retrieve_topology(database = "kegg", species = "hsapiens", 
                                    keyword = c("metabolism", "estrogen"))
names(gsTopology_sub)
```

In addition to working with a single database, users can also choose multiple databases for more comprehensive analyses, albeit with an increased analytic and download time.

```{r dont-run2, eval=FALSE}
gsTopology_2databases <- retrieve_topology(database = c("kegg", "wiki"), 
                                           species = "hs")
```

## Score Single-Sample Pathway Perturbation 

To compute the single-sample expression changes (i.e. logFC) needed for perturbation scores, samples must be in matched pairs. 
For instance, biopsies derived pre- vs post-treatment or untreated vs treated cell lines.
The factor defining the paired structure needs to be passed to the `weight_ss_fc()` function through the `groupBy` parameter. 
In our example dataset, pre- and post-treatment samples are matched by patient IDs. 
Additionally, the sample metadata must include the treatment of all samples. The treatment column must be a factor with the reference level set to be the control treatment. 

```{r}
weightedFC <- weight_ss_fc(as.matrix(logCPM), metadata = sample_meta,
 sampleColumn = "sample", groupBy = "patient_id", treatColumn = "treatment")
glimpse(weightedFC)
```

The output of `weight_ss_fc()` is a `list` where one element is a matrix of weighted single-sample logFCs ($\delta_{ghi}^*$), with rows corresponding to genes and columns to samples, and the other element is the vector of gene-wise weights ($w_g$) used to calculate the weighted logFC ($\delta_{ghi}^*$), as described above.

The matrix of $\delta_{ghi}^*$ values are then passed to pathway topologies to compute the gene-wise perturbation scores for all genes included in a pathway, before being summed into a single score for each pathway. 
The string `ENTREZID:` was also added to all row names of the $\delta_{ghi}^*$ matrix to be compatible with the format pathway topologies were retrieved in. 
These gene-wise perturbation scores can also be used in downstream analysis to identify genes playing the most significant roles in each pathway, as will be demonstrated in the visualisation section below.
The pathway-level perturbation scores ($S_{hip}$) are returned as a data.frame containing sample and gene-set names. 
In the following steps, we first calculate the gene-level contributions to each pathway ($S_{ghip}$) using the function `raw_gene_pert()` and then obtain pathway-level summaries using `pathway_pert()`.
Pathways with zero perturbation scores across all genes and samples will be dropped at this stage. 

```{r}
genePertScore <- raw_gene_pert(weightedFC$logFC, gsTopology)
ssPertScore <- pathway_pert(genePertScore)
head(ssPertScore)
```

## Sample Permutation for Normalisation and Significance Testing

The values obtained from each pathway will vary greatly due to the variability in topology structures.
To determine the significance of individual scores and transform scores to ensure they are comparable across pathways, sSNAPPY utilises a sample-permutation strategy to simulate the null distributions of perturbation scores. 
Since sample labels will be permuted randomly to put samples into pairs, sample metadata is not required by the `generate_permuted _scores()` function.
All the possible random pairs between samples will be constructed by default, unless otherwise specified by the `NB` parameter. 
In this example dataset with a total of 22 samples, 462 (i.e. 22 $\times$ 21) permuted scores will be computed for each pathway., 

```{r calc-permuted-score, eval=FALSE}
set.seed(123)
permutedScore <- generate_permuted_scores(
    as.matrix(logCPM), gsTopology = gsTopology, weight = weightedFC$weight
)
```

```{r, echo=FALSE}
permutedScore <- readRDS(here::here("data/permutedScore.rds"))
```

Apart from pathways whose permuted perturbation scores are all zero, the rest of the pathways' empirical distributions should be approximately normally distributed with means equal to zero. 
To demonstrate that, we randomly selected 6 pathways and visualised the permuted perturbation scores as boxplots (Figure \@ref(fig:Figure3)).

```{r Figure3, out.width='100%', fig.width=10, fig.height=6, fig.cap="Permuted perturbation scores of six randomly selected pathways. All sampled empirical distributions are approximately normally distributed with a mean of zero."}
set.seed(234)
permutedScore %>%
    keep(~all(.!=0)) %>%
    .[sample(seq_along(.), 6)] %>%
    as.data.frame() %>%
    pivot_longer(
      cols = everything(), names_to = "gs_name", values_to = "score"
    ) %>%
    mutate(
        gs_name = str_replace_all(gs_name, "\\.", " "),
        gs_name = str_remove_all(gs_name, "kegg ")
    ) %>%
    ggplot(aes(gs_name, score, fill = gs_name)) +
    geom_boxplot() +
    scale_x_discrete(labels = function(x) str_wrap(x, width = 10)) +
    scale_fill_discrete(name = "Gene-set Name") + 
    labs(x = "Pathway", y = "Permuted Perturbation Score") +
    theme_bw() + 
    theme(
      legend.position = "none", 
      axis.title = element_text(size = 16), 
      axis.text = element_text(size = 14)
    )
```

Permuted distributions are used to convert each pathway-level score into a scaled robust $Z$-score using the function `normalise_by_permu()`.
Robust Z-scores can in turn be transformed into two-sided p-values and corrected for multiple testing using any of the available methods, and returning the FDR adjusted values by default.
In our brief example data, no pathways would be considered as significantly perturbed at the single-sample level using an FDR adjustment with $\alpha$ = 0.05.


```{r}
normalisedScores <- normalise_by_permu(permutedScore, ssPertScore, 
                                       sortBy = "pvalue")
head(normalisedScores)
```

A key question of interest in this dataset is to identify which biological processes were impacted by chemotherapy across the entire group of patients.
Using the sample-level output obtained above, we can explore this by applying t-tests or regression models across all samples.
In order to minimise spurious results, Smyth's moderated t-statistics[@Smyth_2004] are able to be applied across the complete dataset.
Given that we have robust Z-score, a constant variance was assumed across all pathways.
To perform this analysis, the robust Z-scores were converted to a matrix, and the standard limma methodologies was used with variance set to be constant.
For our use case here, no design matrix is required and a simple t-test is appropriate.


```{r}
score_matrix <- normalisedScores %>%
  dplyr::select(score, gs_name, sample) %>%
  pivot_wider(names_from = "sample", values_from = "score") %>%
  column_to_rownames("gs_name") %>%
  as.matrix()
score_fits <- lmFit(score_matrix, design = rep(1, ncol(score_matrix))) %>% 
  eBayes(trend = FALSE)
top_table <- topTable(score_fits, number = Inf) %>%
  as_tibble(rownames = "gs_name")
```

Pathways with an FDR < 0.05 in the moderated t-test were considered to be significantly perturbed at the group level. 
8 out of the 315 tested KEGG pathways passed the selection threshold (Table \@ref(tab:Table1)).

```{r}
table1 <- top_table %>%
  dplyr::filter(adj.P.Val < 0.05) %>% 
  mutate(
    Direction = ifelse(logFC < 0, "Inhibited", "Activated"), 
    gs_name = str_remove_all(gs_name, "kegg.")
  ) %>%
  dplyr::select(
    Pathway = gs_name, Change = logFC,  P.Value, FDR = adj.P.Val, Direction
  )
```


```{r Table1, echo=FALSE, message=FALSE, warning=FALSE}
knitr::kable(
  table1, format = "latex",
  caption = "Significantly impacted KEGG pathways identified among post-chemotherapy samples using sSNAPPY", 
  booktabs = TRUE,
  format.args = list(digits = 2)
) %>%
  kableExtra::kable_styling(
    latex_options = c("scale_down", "striped")
  ) 
```


For enrichment analysis in the original study[@Zhang2022], unsupervised clustering was performed on all cells labelled as cancer cells. 
Clusters were then annotated manually by performing pathway enrichment testing on cluster marker genes. 
Two clusters, associated with proliferative DNA repair signatures and stress-related markers, contained significantly higher numbers of post-chemotherapy cells than pre-treatment ones[@Zhang2022].
The representative pathways enriched in the stress-associated cluster were *IL6-mediated signaling events*, *TNF signaling pathway*, and *cellular responses to stress*,  characterized by marker genes *JUN*, *FOS*, *IL6*, *TNF*, *CXCR4*, *SNAI1*, *VIM*, *GADD45B*, and *MCL1.* 
Among the stress-related marker genes reported, 5 of them(*FOS*, *GADD45B*, *IL6*, *JUN*, and *TNF*) were implicated in pathways that were considered to be significantly impacted by sSNAPPY.
The other post-chemotherapy cell dominated cluster in the original study was enriched for pathways associated with cell proliferation and DNA repair, such as the Cell cycle, DNA repair, Homology directed repair (HDR) through homologous recombination, and Fanconi anaemia pathway. 
A key gene involved in those pathways was *CHEK1*, which was also found in significantly perturbed pathways detected by sSNAPPY: the p53 signaling pathway and Cellular senescence pathway.

Apart from considering all treated samples as biological replicates, users may elect to perform an analysis incorporating other phenotypic traits that affect patients’ responses to chemotherapy, such as disease stages or tumour grades. 
To do that through the moderated t-statistic strategy and extend the above analysis, all that is required is an appropriate design matrix in the `lmFit` step, or subsetting the samples accordingly

## Visualising Perturbed Pathways as Networks

A valuable feature of sSNAPPY is the provision of various visualisation functions to assist in the interpretation of results. 
Biological pathways are not independent of each other with many genes playing a role across multiple pathways, and as such, visualising pathway analysis results as a network can be a powerful way to intuitively summarise the results and facilitate interpretation of the underlying biology. 
The `plot_gs_network()` function allows users to easily convert a list of perturbed biological pathways to a network where edges between pathway nodes represent overlapping genes. 
Defined by the colorBy parameter, pathway nodes can be coloured by either the pathways’ predicted direction of changes or the significance levels (Figure \@ref(fig:Figure4)). 
The returned plot is a ggplot2 [@Wickham2009] object, meaning that components of the plotting theme and other parameters can be customized as for any other ggplot2 objects.

First we should obtain a subset of pathways to visualise.
In the following example, the most highly ranked pathways were inspected, which involved three steps to prepare the data: 1)  rename the logFC column to reflect the true meaning of the value and, 2) create a categorical variable with the pathway status, 3) transform the p-values for simpler visualisation.

```{r}
sigPathway <- top_table %>%
  dplyr::filter(adj.P.Val < 0.05) %>% 
  dplyr::rename(Z = logFC) %>% 
  mutate(
    status = ifelse(
        Z > 0, "Activated", "Inhibited"
    ),
    status = as.factor(status),
    `-log10(p)` = -log10(P.Value)
  )
```

```{r Figure4, out.width='100%', fig.width = 12, fig.height = 5, fig.cap = "Significantly perturbed KEGG pathways identified among post-chemotherapy samples using sSNAPPY, colored by (A) pathways’ predicted direction of changes and (B) pathways’ -log10(p-values). Pathways with a FDR < 0.05 in the moderated t-test were included."}
set.seed(123)
# Plot the network structure
p1 <- plot_gs_network(
  normalisedScores = sigPathway, gsTopology = gsTopology, colorBy = "status"
) +
  scale_colour_manual(values = c("red", "blue", "grey30")) +
  theme_void()
set.seed(123)
p2 <- plot_gs_network(
  normalisedScores = sigPathway,
  gsTopology = gsTopology, 
  colorBy  = "-log10(p)", 
  color_lg_title = expression(paste(-log[10], "p"))
) +
  scale_colour_viridis_c() +
  theme_void()
p1 + p2 + plot_annotation(tag_levels = "A") 
```


By examining the network structure, we can see that many of the highly connected pathways playing a central role in the network are infection-related (Figure \@ref(fig:Figure4)). 
To summarise related pathways and further enable interpretation, we can apply community detection[@Newman2004] in order to group related pathways into 'communities'.
sSNAPPY’s `plot_community()` function is a one-stop shop for applying a community detection algorithm of the user’s choice to the network structure and annotating identified communities by the most common pathway category, denoting the main biological processes perturbed in that community. 
The most recent categories for KEGG pathways were curated from the [KEGG website](https://www.genome.jp/kegg/pathway.html) and included as part of sSNAPPY. 
Annotation of KEGG pathway communities is automatically completed by calling the in-built data object.
Analyses involving other pathway databases will require user-provided pathway categories. 
In the current dataset, the Louvain method was applied to the network of biological pathways and revealed two primary communities, where one was annotated to be cell cycle and endocrine system related and the other one was most closely related to infectious diseases (Figure \@ref(fig:Figure5)). 

```{r Figure5, out.width='80%', fig.width = 10, fig.height = 6.5,fig.cap = 'Significantly perturbed KEGG pathways identified among post-chemotherapy samples using sSNAPPY, colored by community structures detected through the louvain algorithm. The main biological processes perturbed by the chemo-therapy were shown.'}
set.seed(123)
plot_community(
    normalisedScores = sigPathway,
    gsTopology = gsTopology, 
    colorBy = "status"
    
) +
    scale_colour_manual(values = c("red", "blue")) +
    scale_fill_viridis_d() +
    scale_x_continuous(expand = expansion(0.25)) +
    scale_y_continuous(expand = expansion(0.25)) +
    theme_void() 
```


A key advantage of sSNAPPY is that it does not require the prior identification of differentially expressed genes, as this is not always possible in clinical datasets. 
However, knowing which genes are implicated in the perturbation of pathways, particularly those which influence multiple pathways, can provide valuable insights for hypotheses generation about underlying biological mechanisms. 
Therefore, sSNAPPY provides another visualisation feature called `plot_gs2gene`, which enables the inclusion of select genes from each pathway using network structures.
Users can provide a vector of fold-change estimates to visualise genes within pathways, showing their estimated change in expression. 
As pathways often include hundreds of genes, it is recommended to filter for genes most likely to be playing a significant role.
In this example dataset, only genes within the top 500 when ranking by the magnitude of the mean log fold-change were included (Figure \@ref(fig:Figure6)). 
An alternative strategy will be to select genes based on test-statistics, however, this decision is up to the individual researcher.

```{r}
meanFC <- rowMeans(weightedFC$logFC) / weightedFC$weight
top500 <- rank(1/abs(meanFC)) <= 500
dirFC <- ifelse(meanFC > 0, "Up-Regulated", "Down-Regulated")
```

Since KEGG pathway topologies were retrieved using EntrezIDs, users can provide a data.frame mapping Entrez IDs to their chosen identifiers, such as gene names, through the mapEntrezID parameter to make the visualisations more informative. 
A data.frame converting Entrez IDs to ensemble gene names was derived from the Ensembl Release 101[@Ensembl2022] and has been made available as part of the package. 

```{r}
load(system.file("extdata", "entrez2name.rda", package = "sSNAPPY"))
head(entrez2name)
```

```{r Figure6, out.width='80%', fig.width = 9.6, fig.height = 8, fig.cap = 'Significantly perturbed KEGG pathways identified among post-chemotherapy samples using sSNAPPY, showing any genes in the top 500 ranked for magnitudes of changes in expressions, and which pathways they are likely contributing to. No genes in the top500 were found to contribute to the perturbation of the AMPK signaling pathway or the Toxoplasmosis pathway.'}
set.seed(123)
plot_gs2gene(
    normalisedScores = sigPathway, 
    gsTopology = gsTopology, 
    colorGsBy = "status", 
    mapEntrezID = entrez2name, 
    geneFC = meanFC[top500], 
    edgeAlpha = 1, 
    gsNameSize = 4, 
    gsNodeSize = 4
) +
    scale_colour_gradient2(name = "logFC") +
    scale_fill_manual(values = c("red", "blue", "grey50")) +
    theme_void() 
```

## Identifying Key Gene Contributions

To further investigate a specific pathway and elucidate the key genes that contribute to its perturbation, we can generate a heatmap via `plot_gene_contribution()` which shows the gene-level perturbation scores for the top-ranked members of a given pathway.
This function takes advantage of the plotting capabilities of the `pheatmap` package[@pheatmap], and as such, other annotations are also able to be easily included, such as the disease stage, or which general ranges the pathway-level normalised Z-Scores are in.
Inclusion of the Z-Scores enabled the assessment of the level of perturbation predicted in each sample and key genes involved (Figure \@ref(fig:Figure7)).

```{r Figure7, out.width='100%', fig.width = 12, fig.height = 7, fig.cap = 'Gene-level perturbation scores of the top 20 genes in the \"p53 signalling pathway\" ranked by their average contribution to the perturbation score. Samples were annotated by the stage of cancers, along with robust Z-scores which indicate the extent of perturbation within each sample. The genes ATR and ATM were identified as possible key drivers of the activation of p53 signalling pathway, with downregulation of MDM4 also appearing likely to be playing a role.'}
annotation_df <- normalisedScores %>%
    dplyr::filter(gs_name == "kegg.p53 signaling pathway") %>%
    left_join(dplyr::select(sample_meta, sample, Stage), by = "sample") %>%
    mutate(
      `Z Range` = cut(
        robustZ, breaks = seq(-2, 2, length.out = 6), include.lowest = TRUE
      ), 
      sample = str_remove_all(sample, "_post-NACT")
    ) %>% 
    dplyr::select(sample, `Z Range`, Stage)  
z_levels <- levels(annotation_df$`Z Range`)
annotation_col <- list(
  Stage = c(IIIC = "#4B0055", IVA = "#009B95", IVB = "#FDE333"),
  `Z Range` = setNames(
    colorRampPalette(c("navyblue", "white", "darkred"))(length(z_levels)),
    z_levels
  )
)
plot_gene_contribution(
    genePertMatr = genePertScore$`kegg.p53 signaling pathway` %>%
        set_colnames(str_remove_all(colnames(.), "_post-NACT")),
    color = rev(colorspace::divergex_hcl(100, palette = "RdBu")),
    breaks = seq(-0.0015, 0.0015, length.out = 100), 
    annotation_df =  annotation_df, 
    topGene = 20, filterBy = "mean", 
    mapEntrezID = entrez2name,
    annotation_colors = annotation_col,
    cutree_rows = 3, cutree_cols = 3,
    main = "P53 Signaling Pathway [KEGG]"
)
```


From this heatmap we can identify candidate genes which are likely to be making the biggest contribution to the activation of p53 Signaling Pathway upon chemotherapy, such as *ART* and *ATM*, along with the *MDM4* and members of the GADD45 family. 
Ataxia-telangiectasia mutated (*ATM*) is a well-established oncosuppressor[@Moslemi2021] while haploinsufficient mouse double minute 4 (*MDM4*) is an oncogene that suppresses p53[@Momand]. Mutations of both genes have been observed in many types of cancers[@Choi2016; @Gansmo2016; @Atwal2009]. 
Moreover, the Ataxia telangiectasia and RAD3-related protein kinase (*ATR*) and the growth arrest and DNA damage-inducible gene 45 (*GADD45*) have gained great attention as potential targets for anti-tumor treatments due to their involvement in DNA damage repair[@Liebermann2011; @Hu2022].
ATR inhibitors, in particular, have shown promising results in clinical trials for HGSOC.[@Li2022; @Gong2021]. 

# Discussion

In conclusion, we have presented and provided a demonstration for the R/Bioconductor package *sSNAPPY* which offers a novel single-sample pathway perturbation testing approach, tailored for heterogeneous tissue samples in matched-pair design.
In contrast to many common enrichment methods, sSNAPPY uses pathway topology information to compute perturbation scores which indicate the likely impact on the activity of a pathway, by predicting direction of change.
By applying sSNAPPY to a public scRNA-seq data collected before and after HGSOC patients were subjected to chemotherapy, we demonstrated its ability to detect significant pathway perturbations of various interesting biological processes consistent with, and far beyond what was shown in the original study. 
*sSNAPPY* addresses the limitations of alternative strategies that fail to account for gene-gene interactions encoded by pathway topologies or predict the directionality of pathway activities. 
In addition, the single-sample nature of the method can be utilised to address the increasing demand for personalised medicine. 
Through identifying shared and divergent responses between individuals, *sSNAPPY* can provide valuable insights into the heterogeneous responses across clinical samples.
Overall, we believe sSNAPPY represents a valuable addition to the existing body of pathway analysis methods.

# Data availability

The dataset analysed in this manuscript are stored in the data directory of this GitHub repository. 

# Software availability

* Software available from: https://bioconductor.org/packages/release/bioc/html/sSNAPPY.html 
* Source code available from: https://github.com/Wenjun-Liu/sSNAPPY
* Archived source code at time of publication: [DOI (found on right hand side of a Zenodo record)]
* License: [MIT](https://opensource.org/license/mit/)

# Competing interests 

No competing interests were disclosed

# Grant information 

This work was supported by funding from the National Health and Medical Research Council of Australia (ID 1186647 to W.D. Tilley) and the National Breast Cancer Foundation Australia (ID IIRS-23-069 to W.D. Tilley).

# Acknowledgements

# References
