---
title: "sSNAPPY: a R/Bioconductor package for single-sample directional pathway perturbation analysis"
author: 
  - name: Wenjun Liu
    affiliation: Dame Roma Mitchell Cancer Research Laboratories, Adelaide Medical School, Faculty of Health and Medical Sciences, University of Adelaide, Adelaide, Australia
    email: Corresponding Author wenjun.liu@adelaide.edu.au
  - name: Ville-Petteri Mäkinen
    affiliation: 
    - Australian Centre for Precision Health, Cancer Research Institute, University of South Australia, Adelaide, Australia
    - Computational and Systems Biology Program, Precision Medicine Theme, South Australian Health and Medical Research Institute, Adelaide, Australia
    - Computational Medicine, Faculty of Medicine, University of Oulu, Oulu, Finland
    - Center for Life Course Health Research, Faculty of Medicine, University of Oulu, Oulu, Finland
  - name: Wayne D. Tilley
    affiliation: Dame Roma Mitchell Cancer Research Laboratories, Adelaide Medical School, Faculty of Health and Medical Sciences, University of Adelaide, Adelaide, Australia
  - name: Stephen M. Pederson
    affiliation: 
    - Dame Roma Mitchell Cancer Research Laboratories, Adelaide Medical School, Faculty of Health and Medical Sciences, University of Adelaide, Adelaide, Australia
    - Black Ochre Data Laboratories, Telethon Kids Institute, Adelaide, Australia
    - John Curtin School of Medical Research, Australian National University, Canberra, Australia
abstract: 
 When analysing RNA-Seq data, a common outcome is to detect biological pathways with significantly altered activity between the conditions under investigation. 
 The most common strategies test for over-representation within pre-defined gene-sets for genes showing changed expression[@Subramanian2005-lx; @Young2010-jw], but without accounting for gene-gene interactions encoded by pathway topologies, and not being able to directly predict the directional change of pathway activity. 
 To address these issues, we have deloped a single-sample pathway perturbation analysis method  [*sSNAPPY*](https://bioconductor.org/packages/sSNAPPY), now available as an R/Bioconductor package, which leverages pathway topology information to compute pathway perturbation scores, and predicts the potential direction of change across a set of pathways. 
 Here, we demonstrate the use of *sSNAPPY* by applying the method to public scRNA-seq data, derived from ovarian cancer patient tissues collected before and after chemotherapy. 
 Not only were we able to replicate results reported in the original study, but *sSNAPPY* was also able to detect significant perturbation of other biological processes, yielding far greater insight into the response to treatment.
  *sSNAPPY* represents a novel pathway analysis strategy that takes into consideration of pathway topology to predict impacted biology pathways, both within related samples and across treatment groups. 
  In addition to not relying on the detection of differentially expressed genes, the method and associated R package offer important flexibility and provide powerful visualisation tools.

keywords: RNA-seq, pathway enrichment, R package, topology, KEGG, scRNA-seq
bibliography: sample.bib
output: BiocWorkflowTools::f1000_article

header-includes:
  - \usepackage{endfloat}    
  - \usepackage{booktabs}
  - \usepackage{longtable}
---

**R version**: `r R.version.string`

**Bioconductor version**: `r BiocManager::version()`

<!--  Update the name of the package below to be the name of the workflow package you are working on  
**Package**: `r packageVersion("sSNAPPY")` -->

# Introduction

Using pathway enrichment analysis to gain biological insights from gene expression data is a pivotal step in the analysis and interpretation of RNA-seq data, for which numerous methods have been developed (reviewed in [@Maleki2020-ur; @Mubeen2022-eq]). 
Many existing methods tend to view pathways simply as a collection of gene names, as seen in those relying on the detection of differentially expressed genes and applying over-representation analysis (ORA) strategies, and those scoring all genes using functional class scoring (FCS), such as in Gene Set Enrichment Analysis (GSEA) [@Subramanian2005-lx], arguably the most widely-used approach. 
However, databases such as the Kyoto Encyclopaedia of Genes and Genomes (KEGG)[@OgataKEGGKyotoEncyclopediaa] and WikiPathways[@Martens2021] capture not only which genes are implicated in a certain biological process but also their interactions, activating or inhibitory roles, and their relative importance within the pathway, all of which are overlooked in ORA- and FCS-based approaches. 

To fully utilise that additional information, the latest generation of pathway analysis approaches include many which are topology-based such as SPIA[@Tarca2009-nf], DEGraph[@Jacob2012], NetGSA[@Ma2016] and PRS[@Ibrahim2012], as well as others which explicitly model inter-gene correlations[@Wu2012].
Despite differences in the null hypotheses tested across these approaches, overall, they have demonstrated enhanced sensitivity and specificity due to their abilities to take gene-gene interconnections into account[@Nguyen2019-va; @Ma2019]. 
Nevertheless, most topology-based methods focus only on comparing activities of pathways between two treatment groups and cannot be used to score individual samples. 
However, in heterogenous data where more than one factor may be influencing our observations[@Hanzelmann2013], incorporating scoring within paired samples may be desirable and may be able to reveal more nuanced insights. 
To address this gap, we present a Single-Sample Network and Pathway Perturbation analysis methodology called [*sSNAPPY*](https://bioconductor.org/packages/sSNAPPY), available as an R/Bioconductor package.
<!-- To address this gap, we present a Single-Sample directional Pathway Perturbation analysis methodology called [*sSNAPPY*](https://bioconductor.org/packages/sSNAPPY), available as an R/Bioconductor package.  -->
This article defines how *sSNAPPY* computes changes in gene expression within paired samples, and propagates this through gene-set topologies to predict the perturbation in pathway activities within paired samples, before providing summarised results across an entire dataset which would include more robust levels of biological replication. 
The practical usage of the *sSNAPPY* R/Bioconductor package will be illustrated through the analysis of a public scRNA-seq dataset using pseudo-bulk strategies.


# Methods

## Implementation

*sSNAPPY* is an R package that has been reviewed and published on the open-source bioinformatics software platform [Bioconductor](https://bioconductor.org/packages/release/bioc/html/*sSNAPPY*.html) with all source code available via GitHub.
The methodology itself is topology-based, designed to compute directional, single-sample, pathway perturbation scores in gene expression datasets with a matched-pair, or nested design (eg. samples collected before and after treatment).
This allows for the detection of pathway perturbations within all samples from a treatment group, but also within individual samples.

To run *sSNAPPY*, the only required data is a log-transformed expression matrix (e.g. logCPM) with matching sample metadata describing treatment groups and the nested structure.
<!-- I've generalised this to log-transformed etc. as it should also work with microarray data... -->
It is assumed that all pre-processing has been performed beforehand, such as the exclusion of low-signal genes or normalisation to minimise technical artefacts like GC-bias.
The first step, performed internally by *sSNAPPY*, is to estimate sample-specific log fold-change ($\delta_{ghi} = \mu_{ghi} - \mu_{g0i}$) for a treatment $h$ across all genes $g$ within each sample $i$, by subtracting expression estimates for the baseline samples $\hat\mu_{g0i}$ from those in the treatment group $h$.
Since it has been shown that in RNA-seq data, genes with lower expression tend to have larger variance and larger estimates of change[@Law2014], we utilise a gene-level weighting strategy to de-emphasise logFC estimates for low-abundance genes.
Gene-level weights $w_g$ are obtained in a treatment-agnostic manner by fitting a loess curve through the relationship between observed gene-level variance ($\widehat\sigma_g$) and average signal ($\bar\mu_{g\cdot\cdot}$), and taking the inverse of the loess-predicted variance as the weight $w_g = a / f(\bar\mu_{g\cdot\cdot})$, where $f(\bar\mu_{g\cdot\cdot})$ is the predicted value from the loess curve and the constant $a$ ensures $\sum w_g = 1$.
We then use these weighted estimates of logFC for calculation of all pathway perturbation scores.

*sSNAPPY* was built upon the group-level topology-based scoring algorithm initially proposed in R package SPIA[@Tarca2009] to propagate genes' changes in expression through pathway topologies to compute a perturbation score for each pathway.
By modifying the algorithm to incorporate single-sample, weighted estimates of changes in expression we are able to quantify changes in a pathway within a given sample, and then model these across all samples within a treatment group.
Thus, we define the single-sample perturbation score ($S_{hip}$) for a given pathway $p$, sample $i$ and treatment $h$:


$$
\begin{aligned}
S_{hip} = \sum_{g \in G_p} \lbrack S_{ghip} - \delta_{ghi}^*\rbrack \text{, where} \\
S_{ghip} = \delta_{ghi}^* + \sum_{g' \in U_{gp}} \beta_{gg'p} \frac{S_{g'hip}}{N_{g'p}} 
\end{aligned}
$$
where:

* $G_p$ represents the set of genes in pathway $p$, such that $g \in G_p$
* $S_{ghip}$ is the gene-, treatment- and sample-specific perturbation score for pathway $p$
* $\delta_{ghi}^* = w_g\delta_{ghi}$ is the weighted logFC of gene $g$ as described above
* $U_{gp}$ is the subset of $G_p$ containing only the genes directly upstream of gene $g$
* $\beta_{gg'p}$ is the pair-wise gene-gene interactions[@Tarca2009] encoded by the topology matrix for genes $g$ and $g'$
* $N_{gp}$ is the number of downstream genes from any gene $g$
* $S_{hip}$ is the accumulated pathway perturbation score for pathway $p$ in treatment $h$ for sample $i$

The Bioconductor package graphite[@Sales2012] provides functions that can be used to retrieve pathway topologies from a database and convert topology information to adjacency matrices.
In order to streamline this process we have implemented a convenience function, where users only need to provide the name of the desired database to retrieve all topology information in the format required by the scoring algorithm with the correct type of gene identifiers.
<!-- Do we need to specify which identifier? -->

To scale the single-sample pathway perturbation scores ($S_{hip}$) so they are comparable across pathways and to test for significance of individual scores, null distributions of perturbation scores for each pathway are generated through a sample permutation strategy, retaining the correct correlation structure between genes within a pathway.
With each permutation, column names (i.e. sample labels) for the logCPM matrix are randomly shuffled while the rest of the scoring algorithm remains unchanged.
We recommend users to perform a minimum of 1000 permutations, requiring at least 8 unique samples.
<!-- How did you get this number? I make choose(8, 2) as 28 or am I missing something obvious? -->
Subsequently, the median and median absolute deviation (MAD) of the permuted perturbation scores will be calculated and used to normalise the raw perturbation scores to robust $Z$-scores and obtain associated two-sided *p*-values.
Since the method is single sample-based, the permutation strategy remains applicable regardless of experimental design.

Apart from assessing whether a pathway's activity changed significantly within an individual sample, users may also be interested in detecting changes at the group-level, which can be performed by modelling scores with regression models, incorporating Smyth's moderated $t$-statistics[@Smyth_2004] as implemented in *limma*[@limma_2015].
The single-sample nature of *sSNAPPY*'s pathway perturbation scores is particularly helpful for datasets with complex experimental designs or known confounding factors as these can also be incorporated into the final regression models.


## Operation
The package has been tested on all operating systems, requiring R > 4.2.0, and can be installed using BiocManager as follows. 

```{r eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
BiocManager::install("sSNAPPY")
```


# Use Cases

## Data

To demonstrate the application of *sSNAPPY*, we used pre-processed counts from a publicly available scRNA-seq dataset, retrieved from Gene Expression Omnibus (GEO) with accession code GSE165897. 
This dataset consists of 11 homogeneously treated high-grade serous ovarian cancer (HGSOC) patients, with samples taken before treatment and after chemotherapy[@Zhang2022].
In the original study, cells were classified into epithelial, stromal, and immune cells, and we have chosen to only focuses on epithelial cells as they were what the original study primarily focused on. 
Since *sSNAPPY* was designed for bulk RNA-seq data, counts of epithelial cells from the same samples were first summed into pseudo-bulk profiles, giving rise to a total of 22 samples. 
We considered a gene detectable if we observed >1.5 counts per million in >11 samples out of 22, representing all samples from a complete treatment group. 
11,101 (33.8%) of the 32,847 annotated genes passed this selection criteria and were included for downstream analyses. 
Conditional quantile normalisation[@Hansen2012] was then applied to mitigate potential biases introduced by gene length and GC content. 
The normalised logCPM matrix of the processed dataset and sample metadata can be downloaded from here.
<!-- 
Looks like we're missing the download link! Is it https://github.com/Wenjun-Liu/F1000_sSNAPPY_manuscript/raw/master/data/logCPM.rds? If so, we should  have it as a tsv instead of an RDS object. Same for the metadata.
-->

The following packages are required for this workflow

```{r, message=FALSE}
library(sSNAPPY)
library(tidyverse)
library(magrittr)
library(ggplot2)
library(cowplot)
library(kableExtra)
library(AnnotationHub) 
library(edgeR)
```

To read in the data:
<!-- Change these to tsv files. Peole tendnot to provide public data as rds objects -->

```{r}
logCPM <- readRDS(here::here("data/logCPM.rds"))
sample_meta <- readRDS(here::here("data/sample_meta.rds"))
head(sample_meta)
```


## Data preparation and retrieval pathway topology

To apply sSNAPPY, the rownames of the logCPM matrix must be converted to Entrez IDs. Genes without an Entrez IDs were removed.
<!-- I think we should do this beforehand to simplify the paper, then just add another line in the data preparation section. That would dispense with all of the following chunk, which really has nothing to do with sSNAPPY -->

```{r, message=FALSE}
ah <- AnnotationHub() %>%
  AnnotationHub::subset(rdataclass == "EnsDb") %>%
  AnnotationHub::subset(str_detect(description, "101")) %>%
  AnnotationHub::subset(genome == "GRCh38")
stopifnot(length(ah) == 1)
ensDb <- ah[[1]]
rownames(logCPM) <- mapIds(ensDb,rownames(logCPM), 
                           "ENTREZID", keytype = "GENENAME")
# Remove genes that couldn't be matched to entrez IDs
logCPM <- logCPM[!is.na(rownames(logCPM)),]
```


As important step, pathway topology information needs to be retrieved from a chosen database.
Using KEGG as an example, the retrieved topology information will be stored as a list where each element corresponds to a pathway and the numbers in the matrices encode gene-gene interaction. 

<!-- I also think your argument `database` in the function ` retrieve_topology()` should be set to `database = c("kegg", "pathbank", "wikipathways", "reactome", "panther", "pharmgkb", "smpdb")`.  -->
<!-- Then you add the line `database <- match.arg(database)` at the start of your function.  -->
<!-- This is best practice, but really only works if you want to use one database.  -->
<!-- Does the line .retrieveTopology(database, pathwayName) work with > 1 databses? -->
<!-- It may also be worth adding a species argument species = c("athaliana", "btaurus", "celegans", "cfamiliaris", "dmelanogaster", "drerio", "ecoli", "ggallus", "hsapiens", "mmusculus", "rnorvegicus", "scerevisiae", "sscrofa", "xlaevis") and again using `species <- match.arg(species)` and manually checking the graphite::pathwayDatabases() data.frame that the combination exists. -->
<!-- Again this will only work if we want to restrict things to one species which it looks like you're doing in `.retrieveTopology()` -->


```{r}
gsTopology <- retrieve_topology(database = "kegg")
```


Instead of downloading the topology matrices of all pathways, it is also possible to provide a restricted set of pathways for a targeted analysis. 
Customised weights could be assigned to different types of gene-gene interaction type by providing a named numeric vector. 

<!-- Have you tried using pattern matching for this? -->
<!-- Might be a dumb idea, but it might work... -->

```{r}
# Only retrieve the topology matrices of 3 specific pathways
sub_path <- c(
    "Glycolysis / Gluconeogenesis", "Citrate cycle (TCA cycle)",
    "Pentose phosphate pathway"
)
gsTopology_sub <- retrieve_topology(database = "kegg",  pathwayName = sub_path)
```

## Score single-sample pathway perturbation 

To compute the single-sample expression changes (logFC) needed for perturbation scores, each treated sample must have a matching control sample, with the factor defining the paired structure passed to the `weight_ss_fc()` function. 
In our example dataset, pre- and post-treatment samples are matched by patient IDs. 
Additionally, the sample metadata must include the treatment of all samples, one level of which must be the control level indicated as a parameter of the `weight_ss_fc()` function.

<!-- I'm a bit concerned about this function and think it's a bit lacking. You don't specify which column to look in for the factor, instead hard-wiring in the requirement for a `treatment` column. Alternative approaches can be done pretty easily. I'd also ensure that the 'treatment' column was a factor before being passed to the function (which it isn't above) and which gets rid of the requirement for providing the baseline level.
I'm assuming that this could then produce weightedFC for multiple treatment levels...?
Maybe a design matrix approach may be better?
I'd also specify which column should contain column names from the `logCPM` matrix -->

```{r}
weightedFC <- weight_ss_fc(
    logCPM, sample_meta, factor = "patient_id", 
    control = "treatment-naive"
)
glimpse(weightedFC)
```

<!-- I also just noticed that all Entrez IDs had 'ENTREZID` placed at the start. -->
<!-- This is quite unconventional and I'd really discourage this. -->
<!-- Just use the IDs as is. Unless the KEGG pathways make you do this...? -->

The output of `weight_ss_fc()` is a `list` where one element is a matrix of single-sample logFCs ($\delta_{ghi}$), with rows corresponding to genes and columns to samples, and the other element is the vector of gene-wise weights ($w_g$) used to calculate the weighted logFC ($\delta_{ghi}^*$), as described above.
<!-- Again, I'm a bit lost. Are the single sample logFCs weighted or raw? -->
<!-- The text read like they were raw so that's why I put that there, but I may have been wrong -->
<!-- I might have it wrong in the following paragraph too. -->

The matrix of $\delta_{ghi}$ values are then passed to pathway topologies along with the vector of gene weights, to compute the gene-wise perturbation scores for all genes included in a pathway, before being summed into a single score for each pathway. 
These gene-wise perturbation scores can also be used in downstream analysis to identify genes playing the most significant roles in each pathway, as will be demonstrated in the visualisation section below.
The pathway-level perturbation scores ($S_{hip}$) are returned as a data.frame containing sample and gene-set names. 

```{r}
# Compute perturbation scores at the gene-level within
# each pathway each treated sample
genePertScore <- raw_gene_pert(weightedFC$logFC, gsTopology)
# Sum gene-level scores to derive pathway-level scores
ssPertScore <- pathway_pert(genePertScore)
head(ssPertScore)
```

## Sample permutation for normalisation and significance testing

To estimate the significance of individual scores and transform the scores so they are comparable across pathways, sSNAPPY utilises a sample-permutation strategy to simulate the null distributions of perturbation scores. Since sample labels will be permuted randomly, sample metadata is not required by the generate_permuted _scores function, instead, users only need to specify the number of treatment groups in the study, including the control level. Since permutation requires a large amount of computational time and memory, sSNAPPY parallelises this step through functions provided by BiocParallel. Users can choose to customize the parallel back-end or the default one returned by `BiocParallel::bpparam( )` will be used. If the number of samples or the size of the chosen pathway database is large, it is recommended to perform the permutation step on a high-performance computer. 

```{r eval=FALSE}
permutedScore <- generate_permuted_scores(
    logCPM, numOfTreat = 2, NB = 1000, 
    gsTopology = gsTopology, weight = weightedFC$weight
)
```

```{r, echo=FALSE}
permutedScore <- readRDS(here::here("data/permutedScore.rds"))
```

Next, using the function normalise_by_permu, the raw perturbation scores will be normalised to robust z-scores using the median and MAD of permuted scores and further converted to two-sided p-values. Defaulted to using the false-discovery rate, the p-values will be corrected for multiple testings using a user-defined approach. In this case, using an FDR of 0.05 as a cut-off, none of the pathways was considered to be significantly perturbed at the individual sample level. 

```{r}
normalisedScores <- normalise_by_permu(permutedScore, ssPertScore)
head(normalisedScores)
```


A key biological question to answer in this study was what biological processes were impacted by chemotherapy across all patients, the answer to which could be obtained by applying t-tests to normalised scores for each pathway, with the null hypothesis being the mean of normalised perturbation scores equals to 0 for a given pathway. To avoid under-estimation of sample variance, the moderated t-statistics approach proposed by Smyth 200414 was adopted here. Being an empirical Bayes strategy, in moderated t-test, estimated sample variances will be adjusted towards the expected variances estimated by pooling information across all genes/pathways. Performances of the moderated t-statistics were evaluated through simulation studies, and the moderated approach demonstrated more advantageous performance in controlling the false discovery rate. 

```{r}
# Normalised perturbation scores were converted to a matrix, 
# with rows corresponding to pathways and columns to samples 
pert_matrix <- normalisedScores %>%
    dplyr::select(robustZ, gs_name, sample) %>%
    pivot_wider(
        names_from = sample, 
        values_from = robustZ
    ) %>%
    column_to_rownames("gs_name") %>%
as.matrix()
# Linear models were fitted for each pathway. No design matrix was 
# specified as all samples were replicates of the same treatment group
fit_kg <- lmFit(pert_matrix)
# Moderated t-statistics are calculated using the eBayes function
fit_kg <- eBayes(fit_kg, trend = abs(rowMeans(pert_matrix)))
```

Pathways with an FDR < 0.05 in the moderated t-test were considered to be significantly perturbed at the group level. 22 out of the 315 tested KEGG pathways passed the selection threshold (Table \@ref(tab:Table1)).

```{r}
# Use topTable to summarise the linear model fit and correct the p-values with FDR
table1 <- topTable(fit_kg,
                   number = Inf) %>%
    rownames_to_column("gs_name") %>%
    mutate(
        logFC = round(logFC, 4),
        gs_name = as.factor(gs_name),
        Direction = ifelse(logFC < 0, "Inhibited", "Activated"),
        Significant = ifelse(adj.P.Val < 0.05, TRUE, FALSE)) %>%
    dplyr::select(
        `Gene-set Name` = gs_name, `Change in Perturbation Score` = logFC, 
        P.Value, FDR = adj.P.Val, Direction, Significant
    )
```


```{r Table1, echo=FALSE, message=FALSE, warning=FALSE}
knitr::kable(table1, format = "latex", caption = "Results of KEGG pathways identified among post-chemotherapy samples using sSNAPPY", 
             booktabs = T) %>%
    kableExtra::kable_styling(
        latex_options = c("scale_down", "striped")
    ) 
```


In the original study[@Zhang2022], to study the effect of chemotherapy, unsupervised clustering was performed on all cells labelled to be cancer cells. Identified cancer clusters were labelled by performing pathway enrichment testing on cluster marker genes. Two clusters, associated with proliferative DNA repair signatures and stress-related markers, were found to contain significantly higher numbers of post-chemotherapy cells then pre-treatment ones (Ta in @Zhang2022). Marker genes reported for those two clusters were observed in pathways that were detected to be significantly perturbed by sSNAPPY(Table \@ref(tab:Table2) ). 

```{r Table2, echo=FALSE, message=FALSE, warning=FALSE, paged.print=FALSE, results="asis"}
table2 <- readxl::read_excel(here::here("sSNAPPY_paper_files/Tables/Table2.xlsx"))
knitr::kable(table2, caption = ". Detection of stress-associated and proliferative DNA repair signatures, as reported by Zhang et al., 2022, in significantly perturbed pathways identified by sSNAPPY.", booktabs = TRUE) %>%
    kableExtra::kable_styling(
        latex_options = c("scale_down", "striped")
    ) 
```


Apart from treating all treated samples as biological replicates, users might also wish to subset samples into groups by phenotypic traits known to affect patients’ responses to chemotherapy, such as the stages of their cancers. To do that through the moderated t-statistic strategy, we simply need to provide a design matrix in the `lmFit` step.

```{r, eval=FALSE}
X <- model.matrix(
    ~0 + Stage, 
    data = sample_meta %>%
        dplyr::filter(treatment == "post-NACT") %>%
        .[match(colnames(pert_matrix), .$patient_id),] %>%
        mutate(
            Stage = ifelse(Stage == "IVA", "IVA", "IIIC/IVB")
        ) )  %>%
    set_colnames(str_remove_all(colnames(.), "Stage")) %>%
    .[,colSums(.) != 0]  
fit_kg2 <- lmFit(pert_matrix, design = X)
```

## Visualise perturbed pathways as networks

sSNAPPY provides various visualisation functions to assist in the interpretation of results. Since biological pathways are not independent of each other and often contain redundant genes, visualising pathway analysis results as a network is a powerful way to not only intuitively summarise the results but also to facilitate the interpretations of the underlying biology. The `plot_gs_network()` function allows users to easily convert a list of significantly perturbed biological pathways to a network where edges between pathway nodes represent overlapping genes. Defined by the colorBy parameter, pathway nodes can be coloured by either the pathways’ predicted direction of changes or the significance levels (Figure \@ref(fig:Figure1)). The returned plot is a ggplot2 [@Wickham2009] object, meaning that components of its theme could be customized just as any other ggplots using the `ggplot2::theme` function. 

```{r, echo=FALSE}
set.seed(123)
```

```{r Figure1, out.width='100%', fig.width = 12, fig.height = 5, fig.cap = "Significantly perturbed KEGG pathways identified among post-chemotherapy samples using sSNAPPY, colored by (A) pathways’ predicted directions of changes and (B) pathways’ -log10(p-values). Pathways with a FDR < 0.05 in the moderated t-test were included."}
# Extract significantly perturbed pathway. To colour the nodes by directions 
# of changes, the column name of the average perturbation scores must be robustZ. 
sigPathway <- topTable(fit_kg, number = Inf) %>% 
    dplyr::filter(adj.P.Val < 0.05) %>%
    rownames_to_column("gs_name") %>%
    dplyr::rename(
        robustZ = AveExpr,
        pvalue = P.Value
    ) 
# Plot the network structure
p1 <- sSNAPPY::plot_gs_network(
    normalisedScores = sigPathway,
    gsTopology = gsTopology, 
    colorBy = "robustZ"
) +
    theme(
        panel.border = element_blank(), 
        panel.background = element_blank()
    )
p2 <- sSNAPPY::plot_gs_network(
    normalisedScores = sigPathway,
    gsTopology = gsTopology, 
    # or color nodes by significance levels
    colorBy  = "pvalue"
) +
    theme(
        panel.border = element_blank(), 
        panel.background = element_blank()
    )
plot_grid(
    p1, p2, 
    labels = c("A", "B")
)
```


By examining the network structure, we can see that many of the highly connected pathways playing a central role in the network are immune-related. To confirm it and further condense the information, we can summarise the network structures into key biological groups by performing community detection. Widely used in network analysis, community detection is a technique used to identify groups of nodes that are more densely connected than to any other nodes in the network[@Newman2004]. sSNAPPY’s `plot_community()` function is a one-stop shop for applying a community detection algorithm of the user’s choice to the network structure and annotating identified communities by the most common pathway category, denoting the main biological processes perturbed in that community. Retrieved directly from the [KEGG website](https://www.genome.jp/kegg/pathway.html), we have curated the most recent categorisations of KEGG pathways and included it as part of the sSNAPPY package.  Annotation of KEGG pathway communities will be automatically completed by calling the in-built database, while analyses involving other pathway databases require user-provided pathway categorisations. 

The defaulted Louvain method was applied to the network of significantly perturbed biological pathways and revealed two community structures, where one was annotated to be endocrine system related and the other one was infectious disease-related (Figure \@ref(fig:Figure2)). 

```{r Figure2, out.width='80%', fig.width = 10, fig.height = 7, fig.cap = 'Significantly perturbed KEGG pathways identified among post-chemotherapy samples using sSNAPPY, colored by (A) pathways’ predicted directions of changes and (B) pathways’ -log10(p-values). Pathways with a FDR < 0.05 in the moderated t-test were included.'}
sSNAPPY::plot_community(
    normalisedScores = sigPathway, 
        gsTopology = gsTopology, 
        colorBy = "robustZ", 
    ) +
    theme(
        panel.border = element_blank(), 
        panel.background = element_blank()
    )
```

Inferred directly from the expression matrix, a key advantage of sSNAPPY is that it does not require the prior definition of differentially expressed genes, which is not always detectable in clinical datasets. However, knowing the genes that are implicated in perturbing biological pathways, particularly those that affect multiple gene-sets, can provide valuable insights for biologists seeking to formulate hypotheses about underlying biological mechanisms. Therefore, sSNAPPY presents another visualisation feature called `plot_gs2gene`, which enables the inclusion of pathway genes in network structures. Although defaulted to colour all gene nodes in grey, users can provide a vector of logFCs to colour the genes by their changes in expression. As pathways are often made of hundreds of genes, it is recommended not to plot all genes included in perturbed pathways but filter for genes more likely to be playing a significant role, achieved through only providing logFCs of genes to plot.  In this example dataset, we chose to only include genes with the top 500 magnitudes of mean logFCs (Figure \@ref(fig:Figure3)). 

```{r}
# Calculate gene-wise mean logFCs
meanFC <- apply(weightedFC$logFC, 1, mean )
# Extract the top 500 meanFCs
top500_FC <- meanFC %>%
    abs() %>%
    sort(decreasing = TRUE, ) %>%
    .[1:500] 
```

Since pathway topologies were retrieved in Entrez IDs, by default, genes’ Entrez IDs will be used to annotate gene nodes in the plot. However, users can provide a data.frame mapping Entrez IDs to their chosen identifiers through the mapRownameTo parameter. A data.frame converting Entrez IDs to ensemble gene names has been made available as part of the package. 

```{r}
# Read in built-in data.frame entrez2name that matches genes’ 
# Entrez IDs to gene names
load(system.file("extdata", "entrez2name.rda", package = "sSNAPPY"))
head(entrez2name)
```

```{r Figure3, out.width='80%', fig.width = 12, fig.height = 10, fig.cap = 'Significantly perturbed KEGG pathways identified among post-chemotherapy samples using sSNAPPY, annotated by community structure identified with the Louvain algorithm.2 communities were formed, both of which were annotated by the pathway category that the majority of the pathways belong to. '}
# Plot the pathway-gene network for genes with top 500 
#magnitudes of mean FCs and label gene nodes by gene names
sSNAPPY::plot_gs2gene(
    normalisedScores = sigPathway, 
    gsTopology = gsTopology, 
    colorGS_By = "robustZ", 
    mapEntrezID = entrez2name, 
    geneFC = top500_FC, 
    edgeAlpha = 0.3, 
    GsName_size = 4
) +
    theme(
        panel.border = element_blank(), 
        panel.background = element_blank()
)
```

## Identify hub genes contributing to pathway perturbation
If we would like to further investigate a specific pathway and elucidate the key genes that contributed to its perturbation, such as the activation of the "p53 signalling pathway," we can employ a heatmap to display the gene-level perturbation scores of all the genes within the pathway and annotate each column (ie. each sample) by the direction of pathway perturbation in that sample or any other sample metadata using the plot_gene_contribution function (Figure \@ref(fig:Figure4)).

```{r Figure4, out.width='100%', fig.width = 18, fig.height = 10, fig.cap = 'Gene-level perturbation scores of all genes included in the \"p53 signalling pathway\" pathway in each sample where columns of samples were annotated by pathway-level perturbation and the stages of cancers. Genes ATR and ATM were the key driver of the activation of p53 signalling pathway. '}
plot_gene_contribution(
    genePertScore = genePertScore, 
    gsToPlot = "p53 signaling pathway", 
    metadata = dplyr::filter(sample_meta, treatment == "post-NACT") %>%
        dplyr::select(sample = patient_id, Stage), 
    annotation_attribute = c("pathwayPertScore", "Stage"), 
    pathwayPertScore = ssPertScore, 
    mapEntrezID = entrez2name
)
```


From this heatmap we can easily identify that the hub genes making the biggest contribution to the activation of p53 signalling pathway upon chemotherapy were gene ART and gene ATM. The Ataxia-telangiectasia mutated (ATM) gene is a well-established oncosuppressor[@Moslemi2021], mutation of which has been observed in many types of cancers[@Choi2016]. Also involved in DNA damage repair, the Ataxia telangiectasia and RAD3-related protein kinase (ATR) gene has been shown to be a promising therapeutic target for HGSOC[@Li2022]. 

# Discussion

In conclusion, the paper showcased an R/Bioconductor package that offers a novel single-sample pathway perturbation testing approach. sSNAPPY utilizes pathway topology information to compute perturbation scores that predict pathways' potential directions of changes in individual samples. This approach addresses the limitations of current strategies that fail to account for gene-gene interactions encoded by pathway topologies or predict the directionality of pathway activities. By applying sSNAPPY to a public scRNA-seq data collected before and after HGSOC patients were subjected to chemotherapy, we demonstrated its ability to detect significant pathway perturbations of various interesting biological processes beyond what were shown in the original studies. Overall, sSNAPPY presents a promising strategy for single sample-based pathway analysis in RNA-seq data.

# Data availability
The dataset analysed in this manuscript are stored in the data directory of this GitHub repository. 

# Software availability

* Software available from: https://bioconductor.org/packages/release/bioc/html/sSNAPPY.html 
* Source code available from: https://github.com/Wenjun-Liu/sSNAPPY
* Archived source code at time of publication: [DOI (found on right hand side of a Zenodo record)]
* License: [MIT](https://opensource.org/license/mit/)

# Competing interests 
No competing interests were disclosed

# Grant information 
Any grants that supported the work must be listed here, including the grant number.  

# Acknowledgements

